node/network/availability-recovery/src/lib.rs

```diff
@@ -0,0 +1,756 @@
+					Err(RecoveryError::Unavailable),
+				)).await.is_err() {
+					return;
+				};
+				break;
+			}
+
+			while self.requesting_chunks.len() < N_PARALLEL {
+				if let Some(validator_index) = self.shuffling.pop() {
+					let (tx, rx) = oneshot::channel();
+
+					if self.to_state.send(FromInteraction::MakeRequest(
+						self.validators[validator_index as usize].clone(),
+						self.candidate_hash.clone(),
+						validator_index,
+						tx,
+					)).await.is_err() {
+						return;
+					};
+
+					self.requesting_chunks.push(rx.timeout(CHUNK_REQUEST_TIMEOUT));
+				} else {
+					break;
+				}
+			}
+
+			// Check if the requesting chunks is not empty not to poll to completion.
+			if !self.requesting_chunks.is_empty() {
+				// Poll for new updates from requesting_chunks.
+				while let Some(request_result) = self.requesting_chunks.next().await {
+					match request_result {
+						Some(Ok(Ok((peer_id, chunk)))) => {
+							// Check merkle proofs of any received chunks, and any failures should
+							// lead to issuance of a FromInteraction::ReportPeer message.
+							if let Ok(anticipated_hash) = branch_hash(
+								&self.erasure_root,
+								&chunk.proof,
+								chunk.index as usize,
+							) {
+								let erasure_chunk_hash = BlakeTwo256::hash(&chunk.chunk);
+
+								if erasure_chunk_hash != anticipated_hash {
+									if self.to_state.send(FromInteraction::ReportPeer(
+											peer_id,
+											COST_MERKLE_PROOF_INVALID,
+									)).await.is_err() {
+										return;
+									}
+								} else {
+									self.received_chunks.insert(peer_id, chunk);
+								}
+							} else {
+								if self.to_state.send(FromInteraction::ReportPeer(
+										peer_id,
+										COST_MERKLE_PROOF_INVALID,
+								)).await.is_err() {
+									return;
+								}
+							}
+						}
+						Some(Err(e)) => {
+							tracing::debug!(
+								target: LOG_TARGET,
+								err = ?e,
+								"A response channel was cacelled while waiting for a chunk",
+							);
+						}
+						Some(Ok(Err(e))) => {
+							tracing::debug!(
+								target: LOG_TARGET,
+								err = ?e,
+								"A chunk request ended with an error",
+							);
+						}
+						None => {
+							tracing::debug!(
+								target: LOG_TARGET,
+								"A chunk request has timed out",
+							)
+						}
+					}
+				}
+			}
+
+			// If received_chunks has more than threshold entries, attempt to recover the data.
+			// If that fails, or a re-encoding of it doesn't match the expected erasure root,
+			// break and issue a FromInteraction::Concluded(RecoveryError::Invalid).
+			// Otherwise, issue a FromInteraction::Concluded(Ok(())).
+			if self.received_chunks.len() >= self.threshold {
+				let concluded = match polkadot_erasure_coding::reconstruct_v1(
+					self.validators.len(),
+					self.received_chunks.values().map(|c| (&c.chunk[..], c.index as usize)),
+				) {
+					Ok(data) => FromInteraction::Concluded(self.candidate_hash.clone(), Ok(data)),
```

Comment:
I don't see where it checks the re-encoding matches the expected erasure root. It would be great to have a unit test for that case as well.

Reply:
Since this loop consists of a few different pieces of functionality, maybe refactor them into separate functions that may be more unit-testable.

Reply:
Yes, rechecking the erasure root calculation is extremely important.  If it fails then the candidate is invalid so we escalate.

Reply:
Yep - escalation will happen elsewhere. This subsystem just needs to return an "invalid" result to be handled